src/python/pants/engine/internals/graph.py

```diff
@@ -865,13 +873,43 @@ async def resolve_dependencies(
             t.address for t in subtargets.subtargets if t.address != request.field.address
         )
 
+    # If the target has `SpecialCasedDependencies`, such as the `archive` target having
+    # `files` and `packages` fields, then we possibly include those too. We don't want to always
+    # include those dependencies because they should often be excluded from the result due to
+    # being handled elsewhere in the calling code.
+    special_cased: Tuple[Address, ...] = ()
```

Comment:
If we added a new rule, we would probably want its result type to be `Addresses` to avoid having to invent more wrapper dataclasse. But, that will result in graph ambiguity to use `await Get(Addresses, Foo)` in this rule, unless we added the new dataclass.

I don't think there's strong enough of motivation yet to factor this out and pay the cost of Yet Another Dataclass and rule. Note that we don't really expect call sites to resolve all their special deps in one single call, without their normal deps included. Instead, we expect them to resolve each distinct field separately, because each field has its own magical handling. `archive` is a good example of this:

https://github.com/pantsbuild/pants/blob/ea542ac164fb73459f7a7a90aff9c6646d2213e6/src/python/pants/core/target_types.py#L295-L304

Reply:
> because each field has its own magical handling

I'm not sure I understand. If this is the case, why aren't we separating the calls for `Dependencies` and `SpecialCasedDependencies` then?

Reply:
Because there are very select places that we want _all_ dependencies, both normal and special cased. Specifically, project introspection needs _all_ dependencies. 

It is simpler in those places to have one call `await Get(Addresses, DependenciesRequest)` than a call for normal deps + call for all the special cased. In particular, this is apparent with TransitiveTargets. With this current design, we don't need to implement a separate transitive targets rule that duplicates the normal rule. It sounds like you tried implementing new distinct dataclasses yesterday and found this to be the case?

My above comment was trying to explain that in almost all other contexts, you do not want _all_ dependencies, both normal and magical, to be in the same result. You want to separate them out. I was pointing to how that it possible with that design.

Reply:
> It sounds like you tried implementing new distinct dataclasses yesterday and found this to be the case?

It would have been extremely trivial if #8542 had been merged, which would have allowed providing a `@union` base as an `@rule` argument. We could have then made a `TransitiveTargetsRequestBase` union, and it would have generated a version of that rule for each type. I have asked you and others to take a look at that PR many times to no avail. Would you consider looking at it again if I updated it?

Reply:
I'll take a look at #8542 and comment there